### DISCUSSION TITLE
Environment secrets and version control for claude_desktop_config.json

### DISCUSSION THREAD

**Discussion by g-simmons** [modelcontextprotocol/.github#108] [Q&A]
## Question
**I am looking for suggestions for how to version my `claude_desktop_config.json` without committing secrets.**

<details><summary>More details</summary>

## Background
Like many of you, I like to version control my dotfiles. I consider `claude_desktop_config.json` to be a dotfile.

My dotfiles are not publicly accessible, but maybe they will be one day. I prefer not to include secrets in my dotfiles.

Some servers that I have seen ([ex](https://github.com/modelcontextprotocol/servers/tree/main/src/github#npx)) require that secrets be included in `claude_desktop_config.json`. To my knowledge, the secrets need to be included verbatim. Please correct me if that's not true. I am getting my info from [this part of the docs.](https://modelcontextprotocol.io/docs/tools/debugging#environment-variables)

## Possible solutions
One solution would be to version control a template, and build `claude_desktop_config.json` from `claude_desktop_config.json.template`. If anyone has already done this I would be curious to see how you did it :)

If there are other ways to get specific variables & values from my default shell environment into my servers, that would be nice to know about :).

## Potentially confusing  part of the docs
The docs state:

> To override the default variables or provide your own, you can specify an `env` key in `claude_desktop_config.json`:
```
{
  "myserver": {
    "command": "mcp-server-myapp",
    "env": {
      "MYAPP_API_KEY": "some_key",
    }
  }
}
```

Someone could interpret this to mean that `"some_key"` can be a _key_, like `"GITHUB_TOKEN"`. That would actually be quite convenient. But example configuration for several servers suggests that it should be a _value_, like an actual GitHub token string. Examples:

Github Server:

```
{
  "mcpServers": {
    "github": {
      "command": "npx",
      "args": [
        "-y",
        "@modelcontextprotocol/server-github"
      ]
    },
    "env": {
      "GITHUB_PERSONAL_ACCESS_TOKEN": "<YOUR_TOKEN>"
    }
  }
  ```
}

Brave Search:

```
{
  "mcpServers": {
    "brave-search": {
      "command": "npx",
      "args": [
        "-y",
        "@modelcontextprotocol/server-brave-search"
      ],
      "env": {
        "BRAVE_API_KEY": "YOUR_API_KEY_HERE"
      }
    }
  }
}
```

I've confirmed that setting `"some_key"` to an environment _key_ does not work with the following test:

in my shell:
```bash
export BRAVE_API_KEY=<my-actual-key-string>
```

claude_desktop_config.json
```json
...
"brave-search": {
      "command": "npx",
      "args": [
        "-y",
        "@modelcontextprotocol/server-brave-search"
      ],
      "env": {
        "BRAVE_API_KEY": "BRAVE_API_KEY"
      }
    },
...
```

Quit & Restart Claude Desktop

Result:
<img width="924" alt="image" src="https://github.com/user-attachments/assets/0ad89283-fd7e-4664-a82a-ec983aac9636" />

Brave search works again if I set

```json
 "env": {
        "BRAVE_API_KEY": "theactualkeystring"
      }
```
</details>

### Pre-submission Checklist

- [X] I have checked that this question would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions and documentation for answers

### Question Category

- [ ] Protocol Specification
- [ ] SDK Usage
- [ ] Server Implementation
- [ ] General Implementation
- [ ] Documentation
- [X] Other

**Comment by g-simmons** [comment]
I will be doing this for now:
```
#!/bin/bash
cat claude_desktop_config.json.template \
  | sed -E "s/:\s*\"(BRAVE_API_KEY)\"/:\"$BRAVE_API_KEY\"/g" \
  > claude_desktop_config.json
```

But I am still curious to see other answers :)

**Reply by adamavenir** [reply]
You might try 1Password as a pretty good UX option for this. You can do:

```
AWS_ACCESS_KEY_ID="op://development/aws/Access Keys/access_key_id"
AWS_SECRET_ACCESS_KEY="op://development/aws/Access Keys/secret_access_key"
```
and 
```
MYSQL_DATABASE = "op://$APP_ENV/mysql/database"
MYSQL_USERNAME = "op://$APP_ENV/mysql/username"
MYSQL_PASSWORD = "op://$APP_ENV/mysql/password"
```

[See docs here](https://developer.1password.com/docs/cli/secrets-environment-variables/#step-3-differentiate-between-environments)

**Reply by g-simmons** [reply]
Thanks for this, I wasn't aware that 1Password had this level of developer support.

**Comment by pythoninthegrass** [comment]
fwiw I hacked together a solution using `~/.bash_aliases` and `envsubst`:

```bash
# ~/.bash_aliases

# claude
# * NOTE: requires `envsubst` (from `gettext` package)
if [[ -f "$HOME/git/claude_config/.env" ]]; then
    export $(grep -v '^#' "$HOME/git/claude_config/.env" | xargs)
    envsubst < "$HOME/git/claude_config/claude_desktop_config.json.tpl" \
        > "$HOME/git/claude_config/claude_desktop_config.json"
fi
```
then in my git repo `~/git/claude_config` it has an .env file that interpolates variables in `claude_desktop_config.json.tpl` to `claude_desktop_config.json` which are then symlinked to the regular `~/.claude` counterparts.

In the repo my `.gitignore` looks like this:
```text
claude_desktop_config.json
!claude_desktop_config.json.tpl
```
